h2o-core/src/main/java/hex/FoldAssignment.java

```diff
@@ -0,0 +1,95 @@
+package hex;
+
+import water.Iced;
+import water.Key;
+import water.exceptions.H2OIllegalArgumentException;
+import water.fvec.Frame;
+import water.fvec.Vec;
+import water.util.ArrayUtils;
+import water.util.VecUtils;
+
+import java.util.Arrays;
+
+public class FoldAssignment extends Iced<FoldAssignment> {
+    final Vec _fold;
+    final int[] _foldMapping;
+    final boolean _internal;
+
+    FoldAssignment(Vec fold, int[] foldMapping, boolean internal) {
+        _fold = fold;
+        _foldMapping = inverseMapping(foldMapping);
+        _internal = internal;
+    }
+
+    Frame toFrame(Key<Frame> key) {
+        return new Frame(key, new String[]{"fold_assignment"}, new Vec[]{_fold});
+    }
+    
+    void dispose() {
+        if (_internal)
+            _fold.remove();
+        Arrays.fill(_foldMapping, -1);
+    }
+    
+    static FoldAssignment fromUserFoldSpecification(int N, Vec fold) {
+        if ( !fold.isCategorical() && 
+                (!fold.isInt() || 
+                        (!(fold.min() == 0 && fold.max() == N-1) && !(fold.min() == 1 && fold.max() == N)))) // Allow 0 to N-1, or 1 to N
+            throw new H2OIllegalArgumentException("Fold column must be either categorical or contiguous integers from 0..N-1 or 1..N");
+
+        return new FoldAssignment(fold, foldMapping(fold), false);
+    }
+
+    static int nFoldWork(Vec fold) {
+        return foldMapping(fold).length;
```

Comment:
yes, I could just call CollectDomainFast to get actualDomain, and return the length; I don't want to do it because I want to keep the logic in one place the expensive stuff of foldMapping is just the collectdomainfast, the rest does "nothing" in comparison 

so keeping it this way keeps the logic simple

I think I would agree with you if the internal "foldMapping" was ever propagated outside of the class, but it is just internal and makes testing the code easier with 2 methods